### PR TITLE
Add new ca-resource component which fetches declaratively using window.fetch

### PR DESCRIPTION
**Pull request by tapina**
…a remote resource declaratively.

### CODE REVIEW

File: components/ca-resource.html [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+         * <dl>
+         * <dt>ca-resource-fetching</dt><dd>Fired when the resource is about to be fetched, with the fetch options available to be modified.</dd>
+         * <dt>ca-resource-fetched</dt><dd>Fired when the resource has been fetched, with an error or non-error result.</dd>
+         * <dt>ca-resource-ready</dt><dd>Fired when the resource has been fetched successfully. {@link value} will be available.
+         * <dt>ca-resource-unauthorized</dt><dd>Fired when the resource fetch failed with 401 Unauthorised because the user's credentials could not be verified.</dd>
+         * <dt>ca-resource-forbidden</dt><dd>Fired when the resource fetch failed with 403 Forbidden because the user's credentials were not acceptable to retrieve it.</dd>
+         * </dl>
+         * TODO: make this able to PUT resources back by changing the 'value' property
+         */
+        class ApiResourceElement extends HTMLElement {
```
Comment (norisalsaadie) [inline components/ca-resource.html]: The rest of the components seem to start their class names with CA. e.g. https://github.com/cambridgeweblab/common-ui/blob/master/components/ca-select.html#L79 - should this be the case?

File: components/ca-resource.html [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            /**
+             * observedAttributes - get a list of attributes that should effect re-fetch.
+             * @return {Array} list of attributes that effect a re-fetch.
+             */
+            static get observedAttributes() {
+                return ['href'];
+            }
+
+            /**
+             * @description Executes when any attribute is changed on the element
```
Comment (norisalsaadie) [inline components/ca-resource.html]: That's true and false 😄

File: components/ca-resource.html [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+             * @type {Event}
+             * @param {string} attrName - the name of the attribute to have changed
+             * @param {string} oldVal - the old value of the attribute
+             * @param {string} newVal - the new value of the attribute
+             * @returns {undefined}
+             */
+            attributeChangedCallback(attrName, oldVal, newVal) {
+                if (ApiResourceElement.observedAttributes.indexOf(attrName) >= 0) {
```
Comment (norisalsaadie) [inline components/ca-resource.html]: could this use ApiResourceElement.observedAttributes.includes(attrName)?

File: components/ca-resource.html [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            }
+
+            /**
+             * Fetches the resource at the specified {@link href}.
+             * @return {Promise<Response>|null} promise of a Fetch response, or null if the fetch was cancelled
+             */
+            fetch() {
+                if (!document.contains(this)) return null; // Don't fetch if we're not attached to a document.
+
+                this._fetchResponse = this._decodedResponse = undefined;
```
Comment (norisalsaadie) [inline components/ca-resource.html]: Wasn't there a lint rule to error on this type assignment?

File: components/ca-resource.html [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                                this.dispatchEvent(new CustomEvent('ca-resource-ready', {
+                                    detail: decodedResponse,
+                                    bubbles: true
+                                }));
+                            });
+                        } else if (response.status === 401) {
+                            this.dispatchEvent(new CustomEvent('ca-resource-unauthorized', eventInit));
+                        } else if (response.status === 403) {
+                            this.dispatchEvent(new CustomEvent('ca-resource-forbidden', eventInit));
+                        }
```
Comment (norisalsaadie) [inline components/ca-resource.html]: Do we need to handle anything else?

File: components/ca-resource.html (hunk 4 above)
Comment (tapina) [reply components/ca-resource.html]: Nope

File: components/ca-resource.html (hunk 5 above)
Comment (tapina) [reply components/ca-resource.html]: Sometime, probably, but for the future.

File: components/ca-resource.html (hunk 3 above)
Comment (tapina) [reply components/ca-resource.html]: Yep

File: components/ca-resource.html (hunk 2 above)
Comment (tapina) [reply components/ca-resource.html]: Sounds like JavaScript all over.

File: components/ca-resource.html (hunk 1 above)
Comment (tapina) [reply components/ca-resource.html]: Yep, good spot. Copy and paste error.

**Pull request merged**